From 2cbf1248f779766a03d438ad5685b952b67546ac Mon Sep 17 00:00:00 2001
From: Kent Overstreet <kent.overstreet@linux.dev>
Date: Mon, 10 Jul 2023 23:01:51 -0400
Subject: [PATCH 40/46] memory profiling: print negative numbers correctly

We shouldn't see negative numbers, but bugs happen, and this will print
out something more sensible.

Signed-off-by: Kent Overstreet <kent.overstreet@linux.dev>
---
 lib/alloc_tag.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/lib/alloc_tag.c b/lib/alloc_tag.c
index 63af296a5..a17a1696b 100644
--- a/lib/alloc_tag.c
+++ b/lib/alloc_tag.c
@@ -57,11 +57,17 @@ static void allocinfo_stop(struct seq_file *m, void *arg)
 static void alloc_tag_to_text(char *buf, struct codetag *ct)
 {
 	struct alloc_tag *tag = ct_to_alloc_tag(ct);
-	char val[10];
+	s64 bytes = alloc_tag_read(tag);
+	char val[10], *p = val;
 
-	string_get_size(alloc_tag_read(tag), 1,
+	if (bytes < 0) {
+		*p++ = '-';
+		bytes = -bytes;
+	}
+
+	string_get_size(bytes, 1,
 			STRING_SIZE_BASE2|STRING_SIZE_NOSPACE,
-			val, sizeof(val));
+			p, val + ARRAY_SIZE(val) - p);
 
 	buf += sprintf(buf, "%8s ", val);
 	buf += codetag_to_text(buf, ct);
-- 
2.41.0.159.g0bfa463d37

